CMakeLists.txt

```diff
@@ -1,54 +1,102 @@
+set(FOSSILIZE_VULKAN_INCLUDE_PATH "${PROJECT_SOURCE_DIR}/khronos/vulkan"
+  CACHE PATH "include path for vulkan headers")
+set(FOSSILIZE_RAPIDJSON_INCLUDE_PATH "${PROJECT_SOURCE_DIR}/rapidjson/include"
+  CACHE PATH "include path for rapidjson headers")
+
+# TODO: Gate all of this behind a better generator expression
+if (CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    list(APPEND FOSSILIZE_CXX_FLAGS
+      -Wall -Wextra -Wunused-result -Wno-missing-field-initializers -Wno-empty-body -Wshadow
+      $<$<NOT:$<BOOL:${FOSSILIZE_ENABLE_EXCEPTIONS}>>:-fno-exceptions>)
 elseif (MSVC)
-    if (FOSSILIZE_MSVC_ANALYZE)
-        set(FOSSILIZE_ANALYZE_FLAGS "/analyze")
-    else()
-        set(FOSSILIZE_ANALYZE_FLAGS "")
-    endif()
+    # XXX: This can be done, and safely, but will require more work and a more recent version of CMake
     # Disabling exceptions in MSVC without a ton of scary warnings is quite finicky.
     # Don't bother. We can verify no exceptions are present with -fno-exceptions on GCC/Clang.
-    set(FOSSILIZE_CXX_FLAGS ${FOSSILIZE_ANALYZE_FLAGS} /D_CRT_SECURE_NO_WARNINGS /D_SCL_SECURE_NO_WARNINGS /wd4267 /wd4244 /wd4309 /wd4005 /MP /DNOMINMAX)
+    list(APPEND FOSSILIZE_CXX_FLAGS
+      $<$<BOOL:${FOSSILIZE_MSVC_ANALYZE}>:/analyze>
+      /wd4267
+      /wd4244
+      /wd4309
+      /wd4005
+      /MP)
 endif()
 
-option(FOSSILIZE_SANITIZE_ADDRESS "Sanitize address" OFF)
-set(FOSSILIZE_LINK_FLAGS)
+# TODO: This does not currently support the MSVC support for AddressSanitizer
 if (FOSSILIZE_SANITIZE_ADDRESS)
-    set(FOSSILIZE_CXX_FLAGS ${FOSSILIZE_CXX_FLAGS} -fsanitize=address)
-    set(FOSSILIZE_LINK_FLAGS "${FOSSILIZE_LINK_FLAGS} -fsanitize=address")
+    list(APPEND FOSSILIZE_CXX_FLAGS -fsanitize=address)
+    list(APPEND FOSSILIZE_LINK_FLAGS -fsanitize=address)
 endif()
 
-option(FOSSILIZE_SANITIZE_THREADS "Sanitize threads" OFF)
 if (FOSSILIZE_SANITIZE_THREADS)
-    set(FOSSILIZE_CXX_FLAGS ${FOSSILIZE_CXX_FLAGS} -fsanitize=thread)
-    set(FOSSILIZE_LINK_FLAGS "${FOSSILIZE_LINK_FLAGS} -fsanitize=thread")
+    list(APPEND FOSSILIZE_CXX_FLAGS -fsanitize=thread)
+    list(APPEND FOSSILIZE_LINK_FLAGS -fsanitize=thread)
 endif()
 
-include(GNUInstallDirs)
-
-add_library(miniz STATIC miniz/miniz.h miniz/miniz.c)
-set_target_properties(miniz PROPERTIES POSITION_INDEPENDENT_CODE ON)
-target_include_directories(miniz PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/miniz)
-if (CMAKE_COMPILER_IS_GNUCXX OR (${CMAKE_CXX_COMPILER_ID} MATCHES "Clang"))
-    target_compile_options(miniz PRIVATE -fvisibility=hidden)
-    if (NOT APPLE AND NOT ANDROID)
-        target_compile_definitions(miniz PRIVATE _LARGEFILE64_SOURCE)
-    endif()
+if (NOT TARGET miniz)
+  add_library(miniz STATIC miniz/miniz.h miniz/miniz.c)
+  set_target_properties(miniz PROPERTIES POSITION_INDEPENDENT_CODE ON)
+  target_include_directories(miniz
+    PUBLIC
+      $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/miniz>)
+  string(CONCAT largefile $<AND:
```

Comment:
This is unreadable.

Reply:
I can break up some of the logic into something like `set(is-gnu-or-clang ...)`, but I personally find it no more readable than lisp. It can be easier to read if a newer CMake minimum version was used, but I'm unsure what that policy is.

Reply:
Minspec is 3.5 as the CMakeLists.txt declares. If it's possible to use some temporary variables here, that would probably make it more readable.

Reply:
Will do in my next commit. 🙂

Reply:
Let me know if the newer more broken up form is a bit easier to understand.

Reply:
It is, I can accept this. However, the -fvisibility=hidden is gone now.

Reply:
The visibillity is now set above on lines 23 and 27 respectively. CMake has a builtin variable to set this flag. 🙂